src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -460,13 +460,40 @@ static CHIP_ERROR GetAddrInfo(void * context, DnssdResolveCallback callback, uin
     protocol = kDNSServiceProtocol_IPv6;
 #endif
 
-    err = DNSServiceGetAddrInfo(&sdRef, 0 /* flags */, interfaceId, protocol, hostname, OnGetAddrInfo, sdCtx);
-    VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err, true), CHIP_ERROR_INTERNAL);
+    if (interfaceId != static_cast<uint32_t>(-1))
+    {
+        // -1 is the local only interface. If we're not on that interface, we need to get the address for the given hostname.
+        err = DNSServiceGetAddrInfo(&sdRef, 0 /* flags */, interfaceId, protocol, hostname, OnGetAddrInfo, sdCtx);
+        VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err, true), CHIP_ERROR_INTERNAL);
 
-    err = DNSServiceSetDispatchQueue(sdRef, chip::DeviceLayer::PlatformMgrImpl().GetWorkQueue());
-    VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err, true), CHIP_ERROR_INTERNAL);
+        err = DNSServiceSetDispatchQueue(sdRef, chip::DeviceLayer::PlatformMgrImpl().GetWorkQueue());
+        VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err, true), CHIP_ERROR_INTERNAL);
 
-    return MdnsContexts::GetInstance().Add(sdCtx, sdRef);
+        return MdnsContexts::GetInstance().Add(sdCtx, sdRef);
+    }
+    else
+    {
+        sockaddr_in6 sockaddr;
+        memset(&sockaddr, 0, sizeof(sockaddr));
+        sockaddr.sin6_len    = sizeof(sockaddr);
+        sockaddr.sin6_family = AF_INET6;
+        sockaddr.sin6_addr   = in6addr_loopback;
+        sockaddr.sin6_port   = htons((unsigned short) port);
+        uint32_t ttl         = 120; // default TTL for records with hostnames is 120 seconds
+        uint32_t interface   = 0;   // Set interface to ANY (0) - network stack can decide how to route this.
+        OnGetAddrInfo(nullptr, 0 /* flags */, interface, kDNSServiceErr_NoError, hostname,
+                      reinterpret_cast<struct sockaddr *>(&sockaddr), ttl, sdCtx);
+
+        // Don't leak memory.
```

Comment:
This is still leaking sdCtx itself, right?

It would probably be better to call `MdnsContexts::Delete(sdCtx)` and if that fails because it has a null `context->serviceRef` we should add a check there?

Also, there are other leaks all over this code on failre paths; I'll deal with those....

Reply:
@harimau-qirex - would you be able to add a delete call in here and ensure it works?

Reply:
I've made the change and tested that it works. I'm unsure how to push it directly, so I've created another pull request with the change: cecille#2

Reply:
The PR to this branch is now merged, so I'm going to resolve this comment.